pkg/operator/controllers/workaround/systemreserved.go

```diff
@@ -0,0 +1,197 @@
+		"Standard_D8s_v3":   {"MemGB": "32", "vCPUs": "8"},
+		"Standard_E16s_v3":  {"MemGB": "128", "vCPUs": "16"},
+		"Standard_E32s_v3":  {"MemGB": "256", "vCPUs": "32"},
+		"Standard_E4s_v3":   {"MemGB": "32", "vCPUs": "4"},
+		"Standard_E8s_v3":   {"MemGB": "64", "vCPUs": "8"},
+		"Standard_F16s_v2":  {"MemGB": "32", "vCPUs": "16"},
+		"Standard_F32s_v2":  {"MemGB": "64", "vCPUs": "32"},
+		"Standard_F4s_v2":   {"MemGB": "8", "vCPUs": "4"},
+		"Standard_F8s_v2":   {"MemGB": "16", "vCPUs": "8"},
+	}
+	// convert total memory on the VM to amount to reserve
+	memReserved = map[string]string{
+		"8":   "1.8",
+		"16":  "2.6",
+		"32":  "3.56",
+		"64":  "5.48",
+		"128": "9.32",
+		"256": "11.88",
+	}
+	// convert vCPUs to millicores to reserve
+	cpuReserved = map[string]string{
+		"2":  "100",
+		"4":  "140",
+		"8":  "180",
+		"16": "260",
+		"32": "420",
+		"64": "740",
+	}
+)
+
+func NewSystemReserved(log *logrus.Entry, mcocli mcoclient.Interface, clustercli clusterapi.Interface, dh dynamichelper.DynamicHelper) *systemreserved {
+	utilruntime.Must(mcv1.AddToScheme(scheme.Scheme))
+	return &systemreserved{
+		mcocli:     mcocli,
+		clustercli: clustercli,
+		dh:         dh,
+		log:        log,
+	}
+}
+
+func (sr *systemreserved) IsRequired(clusterVersion *version.Version) bool {
+	// check version - currently all versions, but when the upper limit is know
+	// set this.
+
+	return true
+}
+
+func (sr *systemreserved) kubeletConfig() (*runtime.RawExtension, error) {
+	vmSize, err := sr.vmSize()
+	if err != nil {
+		return nil, err
+	}
+	cap, ok := vmCapacity[vmSize]
+	if !ok {
+		return nil, fmt.Errorf("vmSize %s not valid", vmSize)
+	}
+
+	kubeletConfig := map[string]map[string]string{
+		"systemReserved": {
+			"memory": memReserved[cap["MemGB"]] + "G",
+			"cpu":    cpuReserved[cap["vCPUs"]],
+		},
+	}
+	cfgJSON, err := json.Marshal(kubeletConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &runtime.RawExtension{Raw: cfgJSON}, nil
+}
+
+func (sr *systemreserved) vmSize() (string, error) {
+	machinesets, err := sr.clustercli.MachineV1beta1().MachineSets(workerMachineSetsNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return "", err
+	}
+	for _, machineset := range machinesets.Items {
+		if machineset.Spec.Template.Spec.ProviderSpec.Value == nil {
+			return "", fmt.Errorf("provider spec is missing in the machine set %q", machineset.Name)
+		}
+
+		o, _, err := scheme.Codecs.UniversalDeserializer().Decode(machineset.Spec.Template.Spec.ProviderSpec.Value.Raw, nil, nil)
+		if err != nil {
+			return "", err
+		}
+
+		machineProviderSpec, ok := o.(*azureproviderv1beta1.AzureMachineProviderSpec)
+		if !ok {
+			// This should never happen: codecs uses scheme that has only one registered type
+			// and if something is wrong with the provider spec - decoding should fail
+			return "", fmt.Errorf("failed to read provider spec from the machine set %q: %T", machineset.Name, o)
+		}
+
+		return machineProviderSpec.VMSize, nil
```

Comment:
Does this just return the first valid VMSize from the machine sets? We have definitely seen customers where they have machinesets of differing sizes, do we want to use the maximum value?